srml/support/procedural/src/storage/transfo.rs

```diff
@@ -0,0 +1,619 @@
+// Copyright 2017-2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+// tag::description[]
+//! `decl_storage` macro transformation
+// end::description[]
+
+use srml_support_procedural_tools::syn_ext as ext;
+
+use proc_macro::TokenStream;
+
+use syn::{
+	Ident,
+	GenericParam,
+};
+
+use super::*;
+
+pub fn decl_storage_impl(input: TokenStream) -> TokenStream {
+	let def = parse_macro_input!(input as StorageDefinition);
+
+	// old macro naming convention (s replaces $)
+	let StorageDefinition {
+		runtime_crate,
+		visibility,
+		ident: storetype,
+		module_ident,
+		mod_param: strait,
+		crate_ident: cratename,
+		content: ext::Braces { content: storage_lines, ..},
+		extra_genesis,
+		..
+	} = def;
+	let scrate = runtime_crate.map(|rc|rc.ident.content).map(|i|quote!(#i)).unwrap_or_else(||
+		if ::std::env::var("CARGO_PKG_NAME").unwrap() == "srml-support" {
+			quote!( crate )
+		} else {
+			// TODO switch to a more relevant srml_support? or use a rust2018 notation already
+			quote!( ::runtime_support )
```

Comment:
I reuse the component, but also put a big TODO as it is a bad copy paste. It may be an idea to move the tools crate and include sr-api-macros tools.rs in it.